drivers/wifi/esp_at/esp_offload.c

```diff
@@ -405,75 +426,44 @@ static int esp_send(struct net_pkt *pkt,
 	return esp_sendto(pkt, NULL, 0, cb, timeout, user_data);
 }
 
-#define CIPRECVDATA_CMD_MIN_LEN (sizeof("+CIPRECVDATA,L:") - 1)
-#define CIPRECVDATA_CMD_MAX_LEN (sizeof("+CIPRECVDATA,LLLL:") - 1)
-
-static int cmd_ciprecvdata_parse(struct esp_socket *sock,
-				 struct net_buf *buf, uint16_t len,
-				 int *data_offset, int *data_len)
+MODEM_CMD_DEFINE(on_cmd_ciprecvdata)
 {
-	char cmd_buf[CIPRECVDATA_CMD_MAX_LEN + 1];
-	char *endptr;
-	size_t frags_len;
-	size_t match_len;
-
-	frags_len = net_buf_frags_len(buf);
-	if (frags_len < CIPRECVDATA_CMD_MIN_LEN) {
-		return -EAGAIN;
-	}
-
-	match_len = net_buf_linearize(cmd_buf, CIPRECVDATA_CMD_MAX_LEN,
-				      buf, 0, CIPRECVDATA_CMD_MAX_LEN);
-	cmd_buf[match_len] = 0;
-
-	*data_len = strtol(&cmd_buf[len], &endptr, 10);
-	if (endptr == &cmd_buf[len] ||
-	    (*endptr == 0 && match_len >= CIPRECVDATA_CMD_MAX_LEN) ||
-	    *data_len > CIPRECVDATA_MAX_LEN) {
-		LOG_ERR("Invalid cmd: %s", cmd_buf);
-		return -EBADMSG;
-	} else if (*endptr == 0) {
-		return -EAGAIN;
-	} else if (*endptr != _CIPRECVDATA_END) {
-		LOG_ERR("Invalid end of cmd: 0x%02x != 0x%02x", *endptr,
-			_CIPRECVDATA_END);
-		return -EBADMSG;
-	}
+	struct esp_data *dev = CONTAINER_OF(data, struct esp_data,
+					    cmd_handler_data);
+	struct esp_socket *sock = dev->rx_sock;
+	int err;
 
-	/* data_offset is the offset to where the actual data starts */
-	*data_offset = (endptr - cmd_buf) + 1;
+	int socket_data_length = atoi(argv[0]);
 
-	/* FIXME: Inefficient way of waiting for data */
-	if (*data_offset + *data_len > frags_len) {
+	if ((net_buf_frags_len(data->rx_buf) - 2) < socket_data_length) {
+		LOG_DBG("Not enough data -- wait!");
 		return -EAGAIN;
 	}
 
-	*endptr = 0;
-
-	return 0;
-}
+	struct sockaddr_in *recv_addr =
+			(struct sockaddr_in *) &sock->context->remote;
+	int port = atoi(argv[2]);
 
-MODEM_CMD_DIRECT_DEFINE(on_cmd_ciprecvdata)
-{
-	struct esp_data *dev = CONTAINER_OF(data, struct esp_data,
-					    cmd_handler_data);
-	struct esp_socket *sock = dev->rx_sock;
-	int data_offset, data_len;
-	int err;
+	recv_addr->sin_port = ntohs(port);
+	recv_addr->sin_family = AF_INET;
+	/* IP addr comes within quotation marks, which is disliked by
+	 *  conv function. So we remove them.
+	 */
+	char remote_ip_addr[INET_ADDRSTRLEN];
+	size_t remote_ip_str_len = strlen(argv[1]);
 
-	err = cmd_ciprecvdata_parse(sock, data->rx_buf, len, &data_offset,
-				    &data_len);
-	if (err) {
-		if (err == -EAGAIN) {
-			return -EAGAIN;
-		}
+	strncpy(remote_ip_addr, &argv[1][1], remote_ip_str_len - 2);
+	remote_ip_addr[remote_ip_str_len - 2] = '\0';
```

Comment:
I do not quite follow the logic here, but guessing that -2 is there for the quotation marks? Perhaps a more clear comment could be added for that.
When copying strings, we should figure out the limit from the copy target, and not from the source.
So I suggest:
```
	remote_ip_str_len = MIN(sizeof(remote_ip_addr) - 1, strlen(argv[1]) - 2); 
	strncpy(remote_ip_addr, &argv[1][1], remote_ip_str_len);
	remote_ip_addr[remote_ip_str_len] = '\0';
```
This is not tested.

Reply:
Added suggested solution and extended comment. Regarding test did you mean if 'remote_ip_str_len' is to big for 'remote_ip_addr'? I that case i added a check for that aswell.

Reply:
I just meant that I wrote the code snippet without actually testing and verifying that it works ok.

Reply:
I see. I have tested it and it works.